pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go

```diff
@@ -123,51 +131,112 @@ func (cs *cesSubscriber) OnUpdate(oldCES, newCES *cilium_v2a1.CiliumEndpointSlic
+	if !exists {
+		log.WithFields(logrus.Fields{
+			"CESName": CESName,
+			"CEPName": CEPName,
+		}).Info("CEP deleted, calling endpointDeleted")
+		cs.epWatcher.endpointDeleted(c)
+	} else {
+		log.WithFields(logrus.Fields{
+			"CESName": CESName,
+			"CEPName": CEPName,
+		}).Info("CEP deleted, other CEP exists, calling endpointUpdated")
+		cs.epWatcher.endpointUpdated(c, cep)
+	}
+}
+
+// addCEPwithCES insert CEP with CES to the map and triggers endpointUpdated.
+func (cs *cesSubscriber) addCEPwithCES(CEPName, CESName string, newCep *types.CiliumEndpoint) {
+	cs.cepMap.cesMutex.Lock()
+	defer cs.cepMap.cesMutex.Unlock()
+	oldCep, _ := cs.cepMap.getCEPLocked(CEPName)
+	cs.cepMap.insertCEPLocked(CEPName, CESName, newCep)
+	cs.epWatcher.endpointUpdated(oldCep, newCep)
+}
+
+// cepToCESmap is used to map CiliumEndpoint name to CiliumEndpointSlice names.
+// In steady state, there should be exactly one CiliumEndpointSlice associated
+// with a CiliumEndpoint. But when a CEP is being transferred between two CESes,
+// there will be a brief period of time in which the CEP exists in both the CESes.
 type cepToCESmap struct {
+	// cesMutex is used to lock all the operations changing cepMap and ipcache.
 	cesMutex lock.RWMutex
-	cepMap   map[string]string
+	// map of CEP name and map of <CESName, CEP> tuple.
+	cepMap map[string]map[string]*types.CiliumEndpoint
+	// map of CEP name and currently used CES name.
+	// Current CEP is cepMap[CEP][currentCES[CEP]]
+	currentCES map[string]string
 }
 
 func newCEPToCESMap() *cepToCESmap {
 	return &cepToCESmap{
-		cepMap: make(map[string]string),
+		cepMap:     make(map[string]map[string]*types.CiliumEndpoint),
+		currentCES: make(map[string]string),
 	}
 }
 
-func (c *cepToCESmap) insertCEP(cepName, cesName string) {
-	c.cesMutex.Lock()
-	defer c.cesMutex.Unlock()
-	c.cepMap[cepName] = cesName
+func (c *cepToCESmap) insertCEPLocked(cepName, cesName string, cep *types.CiliumEndpoint) {
+	if _, exists := c.cepMap[cepName]; !exists {
+		c.cepMap[cepName] = make(map[string]*types.CiliumEndpoint)
+	}
+	c.cepMap[cepName][cesName] = cep
+	c.currentCES[cepName] = cesName
 }
 
-func (c *cepToCESmap) deleteCEP(cepName string) {
-	c.cesMutex.Lock()
-	defer c.cesMutex.Unlock()
-	delete(c.cepMap, cepName)
+func (c *cepToCESmap) deleteCEPLocked(cepName, cesName string) {
+	cesToCEPMap, exists := c.cepMap[cepName]
+	if !exists {
+		return
+	}
+	if _, exists = cesToCEPMap[cesName]; !exists {
+		return
+	}
+	if len(cesToCEPMap) == 1 {
+		delete(c.cepMap, cepName)
+		delete(c.currentCES, cepName)
+	} else {
+		delete(cesToCEPMap, cesName)
+		if c.currentCES[cepName] == cesName {
+			for k := range cesToCEPMap {
+				c.currentCES[cepName] = k
+				break
+			}
+		}
+	}
 }
 
-func (c *cepToCESmap) getCESName(cepName string) string {
-	c.cesMutex.RLock()
-	defer c.cesMutex.RUnlock()
-	return c.cepMap[cepName]
+// getCEP returns a currently used CEP associated with one of the CESes for the given CEP name.
```

Comment:
getCEPLocked?

Reply:
There's a bit of a convention to name internal functions that don't hold locks as "xxxLocked".

Reply:
Fixed